### DISCUSSION TITLE
Allow automatic returns.

### DISCUSSION THREAD

**Discussion by Shmellyorc** [dotnet/csharplang#7213] [Language Ideas]
Not sure if the title is the exact "term" for it but many languages have built in result. It causes less errors and lower memory footprint. The result would be based on the return value.  THe return value wouldnt even need to called, look at the last example.

Example without built in method result:
```
public int Roll(int dices = 1, int min = 1, int max = 6)
{
    int count = 0;

    for (int i = 0; i < dices; i++)
        count += Random.Integer(min, max + 1);

    return count;
}
```

  
With hidden function built in result:
```
public int Roll(int dices = 1, int min = 1, int max = 6)
{
    for (int i = 0; i < dices; i++)
        result += Random.Integer(min, max + 1);

    return result;
}
```

  
With hidden result, without requesting return. Return would be optional and not required.

Example:
```
public int Roll(int dices = 1, int min = 1, int max = 6)
{
    for (int i = 0; i < dices; i++)
        result += Random.Integer(min, max + 1);
}
```

**Comment by theunrepentantgeek** [comment]
I spent more than a decade programming primarily in Delphi (versions 2-7 for the curious), in which every function had an implicitly declared `Result` variable that merely needed to be assigned somewhere in the body of the function. 

The various teams I worked on all considered it a consistent source of bugs, though they adopted different mitigations. One team found that checking that `Result` was set to the right value on every possible path through a function was a key check to do in any code review - a check that routinely caught edge case bugs.

While there has been a lot of new technology in the interveening 25 years, I doubt that developer brains have been upgraded very much (if at all), and would therefore expect this to be just error prone.

It's also worth noting that there's a lot of code out there already that declares local variables called `result`, raising the potential for this to be a breaking change.

(As an aside: I've reviewed code that used the variable name `reslut`, because the original developer mistyped the name and it was never fixed. Currently anyone typing the right name `result` will get a compilation error. Having a silently declared `result` variable would turn that into a silent failure. One could argue that such code shouldn't exist, but we don't live in that ideal world.)

**Reply by HaloFour** [reply]
VB also has this functionality, and it was the only mechanism for returning values from functions in Classic VB prior to the `Return` keyword being added.  My experience is very similar in that it was a very common source of bugs both because it was very easy to miss the assignment in one branch, but also because it's very easy to accidentally allow your branch to continue and reassign to an incorrect value later.  The explicitness of `return` avoids both potential issues, and I'd much prefer the clarity over any savings that a feature like this might offer.


```vb
Public Function Foo(ByVal s As String) As Integer
    If Len(s) = 0 Then
        Foo = 0
        ' oops, forgot Exit Function
    End If

    ' Do some stuff
    Foo = 1
End Function
```

**Comment by HaloFour** [comment]
>  It causes less errors and lower memory footprint.

My experience is that it causes more errors due to the implicit nature of the code, particularly around branching.  I don't understand the claim about the memory footprint as that is patently not true, as _every_ function now has to declare this implicit local on the stack which may or may not be used.  At most the compiler could elide that local if it is not used, but that just puts it on par with normal code that doesn't rely on an "implicit return".

**Reply by huoyaoyuan** [reply]
In other words, a variable in source code **never** equals to memory space at run time. The compiler only allocates stack memory space when it has to, and will use registers when it can.

**Comment by erik-kallen** [comment]
> It causes less errors

That could be argued, but I would be surprised if it is true.

> and lower memory footprint.

No way.

As a related feature, though, I think it would make sense that if you type `var result = new `, autocompletion would suggest the type returned from the function.

**Reply by Joe4evr** [reply]
> As a related feature, though, I think it would make sense that if you type `var result = new `, autocompletion would suggest the type returned from the function.

I'm pretty sure this already exists, since I've seen it being suggested in VS multiple times recently. No Resharper or other plugins here.
